src/managers/auth.py

```diff
@@ -25,85 +27,148 @@ class Acl:
+
+    The class will load authfile.json to create an internal mapping, after this, operations are
+    only done internally. To persist changes, `write_authfile` has to be invoked e.g.:
+
+    ```
+        auth = KarapaceAuth(context, workload)
+        auth.add_user(username, password)
+        auth.add_acl()
+
+        auth.write_authfile()
+    ```
+    """
 
     def __init__(self, context: ClusterContext, workload: WorkloadBase):
         self.context = context
         self.workload = workload
 
-    @property
-    def parsed_authfile(self) -> dict:
-        """Return authfile parsed as a dict."""
-        return json.loads("\n".join(self.workload.read(self.workload.paths.registry_authfile)))
+        # Load current users and ACLs to have an internal mapping of auth state.
+        self._load_authfile()
 
-    @property
-    def users(self) -> list[str]:
-        """Return existing users on karapace authfile."""
-        full_users = self.parsed_authfile.get("users", [])
-        return [user["username"] for user in full_users]
+    def _load_authfile(self) -> None:
+        """Load current Karapace authfile."""
+        # Internal state of auth to the class
+        self.auth_dict: dict[str, AuthDictEntry] = {}
 
-    @property
-    def permissions(self) -> list[dict[str, str]] | None:
-        """Return existing permissions on karapace authfile."""
-        return self.parsed_authfile.get("permissions")
+        raw_file = self.workload.read(self.workload.paths.registry_authfile)
+        if not raw_file:
+            return
+        authfile = json.loads("\n".join(raw_file))
+
+        users = [UserCredentials(**user) for user in authfile["users"]]
+        acls = [Acl(**acl) for acl in authfile["permissions"]]
+
+        for user in users:
+            self.auth_dict[user.username] = AuthDictEntry(
+                credentials=user,
+                acls=[acl for acl in acls if acl.username == user.username],
+            )
 
     def add_user(self, username: str, password: str, replace: bool = False) -> None:
         """Create a user for Karapace."""
-        if username in self.users and not replace:
-            return logger.info(f"User {username} already exists, skipping creation")
-
-        user_credentials = json.loads(self.workload.mkpasswd(username=username, password=password))
+        if username in self.auth_dict and not replace:
+            logger.info(f"User {username} already exists, skipping creation")
+            return
+
+        user_creds = UserCredentials(
+            **json.loads(self.workload.mkpasswd(username=username, password=password))
+        )
+        # If username doesn't exist we need to initialize the internal dictionary.
+        # Maybe default dict is a more elegant solution here.
+        if username not in self.auth_dict:
+            self.auth_dict[username] = AuthDictEntry(credentials=user_creds, acls=[])
+        else:
+            self.auth_dict[username]["credentials"] = user_creds
```

Comment:
**suggestion:** I think here you can use dict merging, creating an `AuthDictEntry` either way, and then merging it to `auth_dict`.
You might need to make `AuthDictEntry.acls` optional here, but you'd end up with something like:

```python
self.auth_dict = {user: {"credentials": existing_creds, "acls": existing_acls}}
new_entry = {user: {"credentials": new_creds}}

self.auth_dict = self.auth_dict | new_entry

print(self.auth_dict)

>>> {user: {"credentials": new_creds, "acls": existing_acls}}
```